src/pallene/gc.lua

```diff
@@ -28,77 +29,234 @@ local types = require "pallene.types"
 --   2) Identify functions that don't call the GC (directly or indirectly) and don't treat calls to
 --      them as potential GC sites. (Function inlining might mitigate this for small functions)
 --
---   3) Use a flow-based liveliness analysis to precisely identify the commands that a variable
---      appears live at, instead of approximating with first definition and last use.
---
---   4) Use SSA form or some form of reaching definitions analysis so that we we only need to mirror
+--   3) Use SSA form or some form of reaching definitions analysis so that we we only need to mirror
 --      the writes that reach a GC site, instead of always mirroring all writes to a variable if one
 --      of them reaches a GC site.
 
 local gc = {}
 
-function gc.compute_stack_slots(func)
+local function FlowState()
+    return {
+        input  = {},  -- {var_id -> bool?} live variables at block start
+        output = {},  -- {var_id -> bool?} live variables at block end
+        kill   = {},  -- {var_id -> bool?} variables that are killed inside block
+        gen    = {},  -- {var_id -> bool?} variables that become live inside block
+    }
+end
```

Comment:
Another question for a future PR:
1. How similar is our analysis code to the uninitialized analysis
2. Is it worth abstracting some of it to a separate flow analysis file?

Reply:
The code is very similar. I was thinking of having a separate flow analysis file at some point, but wanted to do a couple more cases to help me abstract away the pattern.